azure-pipelines-build-stage-job.yml

```diff
@@ -72,16 +72,25 @@ jobs:
           MANUAL_MM_VERSION: true
         condition: or( eq( variables['Build.Reason'], 'PullRequest' ), eq( variables['Build.SourceBranchName'], 'mm2.1' ), eq( variables['Build.SourceBranchName'], 'dev' ) )
       - bash: |
-          zip upload/mm2-$(COMMIT_HASH)-$(Agent.OS)-Release target/release/mm2 target/release/libmm2.a -j
+          containers=$(docker ps -q | wc -l)
+          echo $containers
+          if [ $containers -gt 0 ]; then
+            docker rm -f $(docker ps -q)
+          fi
+        displayName: 'Clean up Docker containers'
+        # Run unconditionally even if previous steps failed
+        condition: true
```

Comment:
The cleaning seems to be called every time. I guess it will increase the build time since the agent should pull all docker images. What do you think about stopping all docker containers?

Reply:
> What do you think about stopping all docker containers?

This command indeed stops and deletes containers, not images. It's even called "Clean up Docker *containers*". It's not `docker rmi`.

>  I guess it will increase the build time since the agent should pull all docker images

I would also ask you to not guess but double-check - it's quite clear from the build logs that images are not pulled from scratch on repetitive runs: https://dev.azure.com/ortgma/Marketmaker/_build/results?buildId=5422&view=logs&j=7b1b2da3-4abe-585a-6d57-65b1464e2171&t=00342d23-061b-5a73-090c-dcdd2da13341

`2022-02-16T08:49:32.7830750Z Status: Image is up to date for artempikulin/testblockchain:multiarch`

Reply:
Got it, thanks!